util/signing/signer/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+pub mod cryptography;
+
+/// A collection of bytes.
+#[derive(Debug, Clone)]
+pub struct Bytes(pub Vec<u8>);
+
+/// A signature.
+#[derive(Debug, Clone)]
+pub struct Signature(pub Bytes);
+
+/// A public key.
+#[derive(Debug, Clone)]
+pub struct PublicKey(pub Bytes);
+
+/// Version of a key.
+/// Default mean the current key.
+#[derive(Debug, Clone, Default)]
+pub struct KeyVersion(pub String);
+
+/// Id that identify a Key.
+#[derive(Debug, Clone)]
+pub struct KeyId(pub String);
+
+/// Errors thrown by Signer
+#[derive(Debug, thiserror::Error)]
+pub enum SignerError {
+	#[error("Error during signing : {0}")]
+	Sign(String),
+	#[error("Error during public key retrieval : {0}")]
+	PublicKey(String),
+	#[error("Error can't decode provided hex data : {0}")]
+	Hex(String),
+	#[error("Signature not found.")]
+	SignatureNotFound,
+	#[error("public key not found.")]
+	PublicKeyNotFound,
+}
+
+#[async_trait::async_trait]
+pub trait SignerOperations<C: cryptography::Curve> {
+	/// Signs some bytes.
+	async fn sign(&self, message: Bytes) -> Result<Signature, SignerError>;
+
+	/// Gets the public key.
+	async fn public_key(&self) -> Result<PublicKey, SignerError>;
+}
+
+pub struct Signer<O, C>
+where
+	O: SignerOperations<C>,
+	C: cryptography::Curve,
+{
+	operations: O,
+	_curve_marker: std::marker::PhantomData<C>,
+}
+
+/// Signer wraps an implementation of [SignerOperations] and provides a simple API for signing and getting the public key.
+impl<O, C> Signer<O, C>
+where
+	O: SignerOperations<C>,
+	C: cryptography::Curve,
+{
+	pub fn new(operations: O) -> Self {
+		Self { operations, _curve_marker: std::marker::PhantomData }
+	}
+
+	/// Signs some bytes.
+	pub async fn sign(&self, message: Bytes) -> Result<Signature, SignerError> {
+		self.operations.sign(message).await
+	}
+
+	/// Gets the public key.
+	pub async fn public_key(&self) -> Result<PublicKey, SignerError> {
+		self.operations.public_key().await
+	}
```

Comment:
1. You don't have to import the SignerOperations trait everywhere you want to use. 
 2. You also now have an API where you can parameterize Signer<O, C> which can thus be used to expressing pair a provider with a curve in a nice way
 3. This also separates implementation versus general usage experience paths of this API. The implementers can change the requires on Signer to include different traits, do more standardized things under the hood, but signer itself can stay the same.

Reply:
> 1. You don't have to import the `SignerOperations` trait everywhere you want to use.

So instead you have to import the `Signer` type.

> 2. You also now have an API where you can parameterize `Signer<O, C>` which can thus be used to expressing pair a provider with a curve in a nice way

As opposed to just having an `O: SignerOperations<C>` parameter? Or even `O: SignerOperations<C1> + SignerOperations<C2>` and be able to use the same object with both curves, though this is an academic use case at the moment.

> 3. This also separates implementation versus general usage experience paths of this API. The implementers can change the requires on `Signer` to include different traits, do more standardized things under the hood, but signer itself can stay the same.

I would need a specific example to see how it's superior to just putting the bounds on the implementer object, which you would have to do anyway if you use `Signer` generically.

Reply:
> `O: SignerOperations<C1> + SignerOperations<C2>` and be able to use the same object with both curves

OK, I can see how binding a specific curve into a helper object can make the API more ergonomic in this case, because otherwise you'd have to disambiguate with fully-qualified call syntax. I will resurrect the struct types in #969.

Reply:
> So instead you have to import the `Signer` type.

Yes, but then I can pass around the `Signer` type and never have to import `SignerOperations` to call `sign` elsewhere.

Reply:
> I would need a specific example to see how it's superior to just putting the bounds on the implementer object, which you would have to do anyway if you use Signer generically.

This is not particularly specific, but say you wanted to add a middleware layer, e.g., to support something like caching. `SignerOperations` stays the same and this middleware gets injected in `Signer` logic.